### PR TITLE
keygen: add CA extensions to self-signed certificates

### PR DESCRIPTION
**Pull request by speidy**
Self-signed certificates lacks the required CA extensions.
MS RDP client for iOS cannot connect because of it. the app is crashing right after TLS handshake.
I think it worked before we added our openssl.conf file.

### CODE REVIEW

File: keygen/openssl.conf [hunk 1]
```diff
@@ -1,4 +1,41 @@
 [req]
 distinguished_name = req_distinguished_name
+x509_extensions = v3_ca	# The extentions to add to the self signed cert
 
 [req_distinguished_name]
+
+[ v3_ca ]
```
Comment (proski) [inline keygen/openssl.conf]: I would prefer that we don't use spaces inside section names. It could be just `[v3_ca]`

File: keygen/openssl.conf [hunk 2]
```diff
@@ -1,4 +1,41 @@
...
+
+# Extensions for a typical CA
+
+
+# PKIX recommendation.
+
```
Comment (proski) [inline keygen/openssl.conf]: Too many empty lines. You can write: "PKIX recommendation - extensions for a typical CA". That would be one line.

File: keygen/openssl.conf [hunk 3]
```diff
@@ -1,4 +1,41 @@
...
+subjectKeyIdentifier=hash
```
Comment (proski) [inline keygen/openssl.conf]: Please use spaces around `=` consistently.

File: keygen/openssl.conf [hunk 4]
```diff
@@ -1,4 +1,41 @@
...
+# This is what PKIX recommends but some broken software chokes on critical
+# extensions.
+#basicConstraints = critical,CA:true
+# So we do this instead.
+basicConstraints = CA:true
+
+# Key usage: this is typical for a CA certificate. However since it will
+# prevent it being used as an test self-signed certificate it is best
+# left out by default.
+# keyUsage = cRLSign, keyCertSign
```
Comment (proski) [inline keygen/openssl.conf]: Let's distinguish comments and commented out parameters. For instance, comments could start with two hashes. Or the commented out parameters should not have a space after hash.

File: keygen/openssl.conf [hunk 5]
```diff
@@ -1,4 +1,41 @@
...
+# nsCertType = sslCA, emailCA
+
+# Include email address in subject alt name: another PKIX recommendation
+# subjectAltName=email:copy
+# Copy issuer details
+# issuerAltName=issuer:copy
+
+# DER hex encoding of an extension: beware experts only!
+# obj=DER:02:03
+# Where 'obj' is a standard or added object
```
Comment (proski) [inline keygen/openssl.conf]: It would be better to put comments above the parameters they describe.

File: keygen/openssl.conf [hunk 6]
```diff
@@ -1,4 +1,35 @@
...
+
+# Some might want this also
+#nsCertType = sslCA, emailCA
+
+# Include email address in subject alt name: another PKIX recommendation
+# subjectAltName=email:copy
+# Copy issuer details
+#issuerAltName=issuer:copy
+
+# DER hex encoding of an extension: beware experts only!
```
Comment (proski) [inline keygen/openssl.conf]: "beware" is not needed, "experts only" is cleat enough

**Comment by speidy** [comment]
no space after hash means an example line. I would like to keep it that way.

File: keygen/openssl.conf [hunk 7]
```diff
@@ -1,4 +1,35 @@
 [req]
 distinguished_name = req_distinguished_name
+x509_extensions = v3_ca	# The extentions to add to the self signed cert
```
Comment (proski) [inline keygen/openssl.conf]: extentions -> extensions
Please avoid tabs. The comment should be above that line, not at the end.

**Comment by proski** [comment]
I've tested the patch with several Android clients. I don't see any difference w.r.t. authentication. Parallels Client (legacy and non-legacy) refuses to work. Other clients work. If any client is fixed by this patch, let's merge it.

**Comment by speidy** [comment]
In my testing, only iOS client was crashing. Parallels clients have another issues, not related to TLS.

**Pull request merged**